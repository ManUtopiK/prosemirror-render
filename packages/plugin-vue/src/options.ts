import type { Component } from "vue";
import type { InjectionKey } from "vue";
import type { ProsemirrorJSONAttributes } from "./prosemirror-json";

export type VueProsemirrorComponent = string | Component;
export type VueProsemirrorComponentAndProperties = [VueProsemirrorComponent, Record<string, unknown>];
export type VueProsemirrorComponentReturns = VueProsemirrorComponent | VueProsemirrorComponentAndProperties;
export type VueProsemirrorComponentFunction = (attributes: ProsemirrorJSONAttributes) => VueProsemirrorComponentReturns;
export type VueProsemirrorComponentOption = string | VueProsemirrorComponentAndProperties | VueProsemirrorComponentFunction;

export type VueProsemirrorComponents = Record<string, VueProsemirrorComponentOption>;

export interface VueProsemirrorOptions {
  /** Map node types to component names */
  components: VueProsemirrorComponents;
}

export const VueProsemirrorOptionsKey: InjectionKey<VueProsemirrorOptions> = Symbol("prosemirror-options");

export const defaultOptions: VueProsemirrorOptions = {
  components: {
    doc: "div",
    heading: ({ level }) => (level ? `h${level}` : "header"),
    paragraph: "p",
    code: "code",
    code_block: "pre",
    link: "a",
    horizontal_rule: "hr",
    hard_break: "br",
    bold: "b",
    strong: "strong",
    italic: "i",
    em: "em",
    strike: "s",
    underline: "u",
    subscript: "sub",
    superscript: "sup",
    highlight: "mark",
    blockquote: "blockquote",
    bullet_list: "ul",
    ordered_list: "ol",
    list_item: "li",
    image: "img",
    table: "table",
    table_row: "tr",
    table_header: "th",
    table_cell: "td",
  },
};
