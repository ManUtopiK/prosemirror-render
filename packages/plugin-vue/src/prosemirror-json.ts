/**
 * JSON scalar values
 */
export type ProsemirrorJSONValue =
  | string
  | number
  | boolean
  | { [x: string]: ProsemirrorJSONValue }
  | ProsemirrorJSONValue[]
  | null
  | undefined;

/**
 * JSON ProseMirror attributes
 */
export type ProsemirrorJSONAttributes = Record<string, ProsemirrorJSONValue>;

/**
 * JSON ProseMirror common properties for Mark & Node
 */
export interface ProsemirrorJSONCommon {
  type: string;
  attrs?: ProsemirrorJSONAttributes;
}

/**
 * JSON ProseMirror Mark
 */
export type ProsemirrorJSONMark = ProsemirrorJSONCommon & {
  [p: string]: unknown;
};

/**
 * JSON ProseMirror Node
 */
export type ProsemirrorJSONNode = ProsemirrorJSONCommon & {
  marks?: ProsemirrorJSONMark[];
  content?: ProsemirrorJSONNode[];
  text?: string;
  [p: string]: unknown;
};
