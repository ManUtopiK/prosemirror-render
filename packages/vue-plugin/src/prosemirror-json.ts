/**
 * JSON scalar values
 */
export type ProsemirrorJSONValue =
  | string
  | number
  | boolean
  | { [name: string]: ProsemirrorJSONValue }
  | ProsemirrorJSONValue[]
  | null
  | undefined;

/**
 * JSON ProseMirror attributes
 */
export type ProsemirrorJSONAttributes = Record<string, ProsemirrorJSONValue>;

/**
 * JSON ProseMirror common properties for Mark & Node
 */
export interface ProsemirrorJSONCommon {
  type: string;
  attrs?: ProsemirrorJSONAttributes;
}

/**
 * JSON ProseMirror Mark
 */
export type ProsemirrorJSONMark = ProsemirrorJSONCommon & {
  [other: string]: unknown;
};

/**
 * JSON ProseMirror Node
 */
export type ProsemirrorJSONNode = ProsemirrorJSONCommon & {
  marks?: ProsemirrorJSONMark[];
  content?: ProsemirrorJSONNode[];
  text?: string;
  [other: string]: unknown;
};
