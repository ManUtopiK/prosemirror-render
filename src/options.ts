import type { Component } from "vue";
import type { InjectionKey } from "vue";

export interface VueProseMirrorOptions {
  /** Map node types to component names */
  typeMap: Record<string, string | Component>;
}

export const VueProseMirrorOptionsKey: InjectionKey<VueProseMirrorOptions> = Symbol("prosemirror-options");

export const defaultOptions: VueProseMirrorOptions = {
  typeMap: {
    doc: "div",
    heading: "h[level]",
    paragraph: "p",
    code: "code",
    code_block: "pre",
    link: "a",
    horizontal_rule: "hr",
    hard_break: "br",
    bold: "b",
    strong: "strong",
    italic: "i",
    em: "em",
    subscript: "sub",
    superscript: "sup",
    blockquote: "blockquote",
    bullet_list: "ul",
    ordered_list: "ol",
    list_item: "li",
    image: "img",
    table: "table",
    table_row: "tr",
    table_header: "th",
    table_cell: "td",
  },
};
