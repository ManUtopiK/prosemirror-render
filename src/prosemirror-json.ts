/**
 * JSON ProseMirror attributes
 */
export type Attrs = Record<string, any>

/**
 * JSON ProseMirror common properties for Mark & Node
 */
export type Common = {
    readonly type: string;
    readonly attrs?: Attrs;
}

/**
 * JSON ProseMirror Mark
 */
export type Mark = Common

/**
 * JSON ProseMirror Node
 */
export type Node = Common & {
    readonly marks?: Mark[];
    readonly content?: Node[];
    readonly text?: string;
}
