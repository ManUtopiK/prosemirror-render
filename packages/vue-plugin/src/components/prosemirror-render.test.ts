// @vitest-environment happy-dom
import { mount } from "@vue/test-utils";
import ProsemirrorRender from "./prosemirror-render";
import { describe, it, expect } from "vitest";
import { ProsemirrorJSONNode } from "../prosemirror-json";
import { VueProsemirrorOptionsKey } from "../options";

describe("component ProsemirrorRender", () => {
  const nodeSimple: ProsemirrorJSONNode = {
    type: "doc",
    attrs: {
      "data-test": "doc",
    },
    content: [
      {
        type: "paragraph",
        attrs: {
          "data-test": "paragraph",
        },
        content: [
          {
            type: "text",
            text: "content",
          },
        ],
      },
    ],
  };
  const vueSimple = mount(ProsemirrorRender, { props: { node: nodeSimple } });

  it("exists", () => {
    expect(ProsemirrorRender).toBeTruthy();
  });

  it("renders simple node", () => {
    expect(vueSimple.get("[data-test=doc]").element.tagName).toBe("DIV");
    expect(vueSimple.get("[data-test=doc]").element.children).toHaveLength(1);

    expect(vueSimple.get("[data-test=paragraph]").element.tagName).toBe("P");
    expect(vueSimple.get("[data-test=paragraph]").text()).toBe("content");

    expect(vueSimple.html()).toMatchSnapshot();
  });

  it("don't pollute DOM with stringified object", () => {
    expect(vueSimple.html()).not.toContain("object Object");
  });

  const nodeSimpleMark: ProsemirrorJSONNode = {
    type: "doc",
    attrs: {
      "data-test": "doc",
    },
    content: [
      {
        type: "paragraph",
        attrs: {
          "data-test": "paragraph",
        },
        content: [
          {
            type: "text",
            marks: [
              {
                type: "bold",
                attrs: {
                  "data-test": "bold",
                },
              },
            ],
            text: "content",
          },
        ],
      },
    ],
  };
  const vueSimpleMark = mount(ProsemirrorRender, { props: { node: nodeSimpleMark } });

  it("renders simple mark", () => {
    expect(vueSimpleMark.get("[data-test=paragraph]").text()).toBe("content");
    expect(vueSimpleMark.get("[data-test=bold]").text()).toBe("content");

    expect(vueSimpleMark.text()).toContain("content");
    expect(vueSimpleMark.html()).toMatchSnapshot();
  });

  const nodeDoubleMark: ProsemirrorJSONNode = {
    type: "doc",
    content: [
      {
        type: "paragraph",
        content: [
          {
            type: "text",
            marks: [
              {
                type: "bold",
                attrs: {
                  "data-test": "bold",
                },
              },
              {
                type: "italic",
                attrs: {
                  "data-test": "italic",
                },
              },
            ],
            text: "content",
          },
        ],
      },
    ],
  };
  const vueDoubleMark = mount(ProsemirrorRender, { props: { node: nodeDoubleMark } });

  it("renders marks in order", () => {
    expect(vueDoubleMark.get("[data-test=bold]").get("[data-test=italic]")).toBeDefined();
    expect(vueDoubleMark.get("[data-test=bold]").get("[data-test=italic]").text()).toBe("content");

    expect(vueDoubleMark.text()).toContain("content");
    expect(vueDoubleMark.html()).toMatchSnapshot();
  });

  const nodeEmpty: ProsemirrorJSONNode = {
    type: "doc",
  };
  const vueEmpty = mount(ProsemirrorRender, { props: { node: nodeEmpty } });

  it("renders an empty doc", () => {
    expect(vueEmpty.html()).toBe("<div></div>");
  });

  const nodeMixedTextNodes: ProsemirrorJSONNode = {
    type: "doc",
    content: [
      {
        type: "heading",
        attrs: { level: 2 },
        content: [{ type: "text", text: "Simple" }],
      },
      {
        type: "paragraph",
        attrs: {
          "data-test": "paragraph",
        },
        content: [
          { type: "text", text: "This is a " },
          { type: "text", marks: [{ type: "strong" }], text: "basic" },
          { type: "text", text: " example." },
        ],
      },
    ],
  };
  const vueMixedTextNodes = mount(ProsemirrorRender, { props: { node: nodeMixedTextNodes } });

  it("renders mixed text and marks", () => {
    expect(vueMixedTextNodes.get("[data-test=paragraph]").text()).toBe("This is a basic example.");
    expect(vueMixedTextNodes.html()).toMatchSnapshot();
  });

  const unsafeScriptDocument: ProsemirrorJSONNode = {
    type: "doc",
    content: [
      {
        type: "script",
        content: [{ type: "text", text: "console.log('unsafe')" }],
      },
    ],
  };
  const vueUnsafeScript = mount(ProsemirrorRender, {
    props: { node: unsafeScriptDocument },
    global: {
      provide: {
        [VueProsemirrorOptionsKey]: {
          types: {
            script: false,
          },
        },
      },
    },
  });

  it("skip types when false", () => {
    expect(vueUnsafeScript.html()).not.toContain("unsafe");
    expect(vueUnsafeScript.html()).toMatchSnapshot();
  });
});
