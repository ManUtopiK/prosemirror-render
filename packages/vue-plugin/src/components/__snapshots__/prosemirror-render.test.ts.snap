// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`component ProsemirrorRender > renders marks in order 1`] = `
"<div>
  <p><b data-test=\\"bold\\"><i data-test=\\"italic\\">content</i></b></p>
</div>"
`;

exports[`component ProsemirrorRender > renders mixed text and marks 1`] = `
"<div>
  <h2>Simple</h2>
  <p data-test=\\"paragraph\\">This is a <strong>basic</strong> example.</p>
</div>"
`;

exports[`component ProsemirrorRender > renders simple mark 1`] = `
"<div data-test=\\"doc\\">
  <p data-test=\\"paragraph\\"><b data-test=\\"bold\\">content</b></p>
</div>"
`;

exports[`component ProsemirrorRender > renders simple node 1`] = `
"<div data-test=\\"doc\\">
  <p data-test=\\"paragraph\\">content</p>
</div>"
`;

exports[`component ProsemirrorRender > skip types when set to false 1`] = `
"<div>
  <!-- prosemirror type 'script' skipped -->
</div>"
`;

exports[`component ProsemirrorRender > skip unknown types when skipUnknown option set 1`] = `
"<div>
  <!-- prosemirror type 'script' skipped -->
</div>"
`;
